thinc/loss.py

```diff
@@ -96,26 +99,44 @@ def convert_truths(self, truths, guesses: Floats2d) -> Tuple[Floats2d, Floats2d]
                             negatives_mask[i] = 0  # type: ignore
                             negatives_mask[i][neg_index] = -1  # type: ignore
                     truths = [self._name_to_i[name] for name in truths]
+            # cast truths as 1d array
             truths = xp.asarray(truths, dtype="i")
+            # get mask for 1d array
             mask = _make_mask(guesses, missing)
         else:
+            # get mask for 2d array
             mask = _make_mask_by_value(truths, guesses, missing_value)
-        if truths.ndim != guesses.ndim:
-            # transform categorical values to one-hot encoding
-            truths = to_categorical(
-                cast(Ints1d, truths),
-                n_classes=guesses.shape[-1],
-                label_smoothing=self.label_smoothing,
+        truths = cast(IntsOrFloats, truths)
+        # Convert 1d truths to 2d and apply smoothing.
+        if truths.ndim == 1:
+            truths = cast(
+                Floats2d,
+                to_categorical(
+                    truths,
+                    n_classes=guesses.shape[-1],
+                    label_smoothing=self.label_smoothing,
+                ),
             )
-        else:
-            if self.label_smoothing:
+        # Check if 2D array is valid and maybe apply smoothing.
+        elif truths.ndim == 2:
+            # Validate if each row sums to 1.
+            if not xp.all(truths.sum(axis=1) == 1):
```

Comment:
I am a bit worried about this part. What if it doesn't sum up to exactly one due to rounding errors that where incurred while computing the truths table?

Reply:
this is a mistake indeed it should just check if its close.